Record whether data streams for logs-*-* exist for logsdb enablement in 9.x

martijnvg opened the pull request:
Add LogsPatternUsageService that records whether there are data streams matching with `logs-*-*` pattern. This is recorded via the new `logsdb.prior_logs_usage` cluster setting. Upon upgrade to 9.x this can be used to determine whether logsdb should be enabled by default if cluster.logsdb.enabled hasn't been set.

The recommended upgrade path to 9.x is always to go to 8.latest. This component will run in clusters with version greater than 8.18.0 but not on 9.0 and newer.

Note that a follow change is required that removes / disables `LogsPatternUsageService` in main. Additionally updating the default of the `LogsDBPlugin.CLUSTER_LOGSDB_ENABLED` setting:

```java
public static final Setting<Boolean> CLUSTER_LOGSDB_ENABLED = Setting.boolSetting(
        "cluster.logsdb.enabled",
        settings -> {
            if (LOGSDB_PRIOR_LOGS_USAGE.exists(settings)) {
                return Boolean.FALSE.toString();
            } else {
                return Boolean.TRUE.toString();
            }
        },
        Setting.Property.Dynamic,
        Setting.Property.NodeScope
    );
```

martijnvg committed:
LogsPatternUsageService

Add LogsPatternUsageService that records whether there are data streams matching with logs-*-* pattern. This is recorded via logsdb.prior_logs_usage cluster setting. Upon upgrade to 9.x this can be used to determine whether logsdb should be enabled by default if cluster.logsdb.enabled hasn't been set.

Commit:
[CI] Auto commit changes from spotless

martijnvg committed:
iter

martijnvg committed:
more tests

Commit:
[CI] Auto commit changes from spotless

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageService.java:
```diff
@@ -0,0 +1,150 @@
+import java.util.function.Supplier;
+
+/**
+ * A component that check in the background whether there are data streams that match log-*-* pattern and if so records this as persistent
```
Nit: checks*

martijnvg committed:
Merge remote-tracking branch 'es/main' into LogsPatternUsageService

martijnvg committed:
iter

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:
```diff
@@ -76,6 +80,19 @@ public Collection<?> createComponents(PluginServices services) {
+
+        var clusterService = services.clusterService();
+        Supplier<Metadata> metadataSupplier = () -> clusterService.state().metadata();
+        var historicLogsUsageService = new LogsPatternUsageService(services.client(), settings, services.threadPool(), metadataSupplier);
```
Consider initializing in the constructor and calling an `init` function here, similar to `LogsdbIndexModeSettingsProvider`.

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsDBPlugin.java:
```diff
@@ -76,6 +80,19 @@ public Collection<?> createComponents(PluginServices services) {
+        Supplier<Metadata> metadataSupplier = () -> clusterService.state().metadata();
+        var historicLogsUsageService = new LogsPatternUsageService(services.client(), settings, services.threadPool(), metadataSupplier);
+        clusterService.addLocalNodeMasterListener(historicLogsUsageService);
+        clusterService.addLifecycleListener(new LifecycleListener() {
```
Add a function in `LogsPatternUsageService` to construct and return a listener?

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageService.java:
```diff
@@ -0,0 +1,150 @@
+ */
+final class LogsPatternUsageService implements LocalNodeMasterListener {
+
+    private static final String LOGS_PATTERN = "logs-*-*";
```
Deduplicate with `LogsdbIndexModeSettingsProvider.LOGS_PATTERN` ?

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageService.java:
```diff
@@ -0,0 +1,150 @@
+    private static final Logger LOGGER = LogManager.getLogger(LogsPatternUsageService.class);
+    static final Setting<TimeValue> USAGE_CHECK_PERIOD = Setting.timeSetting(
+        "logsdb.usage_check.period",
+        new TimeValue(24, TimeUnit.HOURS),
```
Consider adding a comment explaining how this value is picked.

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageService.java:
```diff
@@ -0,0 +1,150 @@
+        new TimeValue(24, TimeUnit.HOURS),
+        Setting.Property.NodeScope
+    );
+    static final Setting<Boolean> LOGSDB_PRIOR_LOGS_USAGE = Setting.boolSetting(
```
Do we want to restrict this as internal? Or, maybe allow updates in case we get it wrong?

At any rate, let's document expectations.

martijnvg replied:
Only index settings can be made internal/private. Not node/ cluster settings.

I think we shouldn't document this setting. If someone sets this setting, then they might just as well configure `cluster.logsdb.enabled`? The advantage of this being public, is that for debugging purposes we can see that a 8.18 cluster is the right state to upgrade.

I will add a comment on top of this setting constant.

kkrik-es commented on x-pack/plugin/logsdb/src/main/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageService.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    void scheduleNext(TimeValue waitTime) {
+        if (isMaster && hasPriorLogsUsage == false) {
```
So, if the master changes, we may miss an update in 24h? Worth documenting, though this seems fine.

martijnvg replied:
I updated this wait time logic a bit via this commit: e0a5c11c4ebabdf782e963a5f566e4ba9612f039

martijnvg committed:
Merge remote-tracking branch 'es/main' into LogsPatternUsageService

kkrik-es commented on x-pack/plugin/logsdb/src/test/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageServiceIntegrationTests.java:
```diff
@@ -0,0 +1,93 @@
+                .actionGet();
+            assertThat(response.persistentSettings().get("logsdb.prior_logs_usage"), equalTo("true"));
+        });
+    }
```
Add a test for a non-matching pattern? 

Also, is it possible to test that it only applies to master and it no longer runs after setting to `true`?

kkrik-es reviewed (commented):
Looks good, just a few minor comments.

martijnvg committed:
Start checking for logs-*-* usage quicker after master node has been elected.
Changed `logsdb.usage_check.period` to `logsdb.usage_check.max_period` and
starting to check 60s after node has been elected as master. This gets doubled
after each check that doesn't see any log-*-* usage upto `logsdb.usage_check.max_period` (defaults to 24h)

martijnvg committed:
applied feedback

martijnvg committed:
verify that `nextWaitTime` gets double after no log-*-* usage has been found

kkrik-es commented on x-pack/plugin/logsdb/src/test/java/org/elasticsearch/xpack/logsdb/LogsPatternUsageServiceIntegrationTests.java:
```diff
@@ -73,6 +79,31 @@ public void testLogsPatternUsage() throws Exception {
+        assertThat(indexResponse.getResult(), equalTo(DocWriteResponse.Result.CREATED));
+
+        ensureGreen("log-myapp-prod");
+        assertBusy(() -> {
```
In fairness, add a short sleep() to give time for the usage service to run?

martijnvg replied:
Good point. I did this which is better than a thread.sleep() :572d670010d88ff1f1639197fe18fec7544ddf88

martijnvg committed:
give LogsPatternUsageService opportunity to run

elasticmachine committed:
Merge branch 'main' into LogsPatternUsageService

martijnvg commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into LogsPatternUsageService

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 120708`
